src/linux/mod.rs

```diff
@@ -44,61 +46,165 @@ impl Sandbox for LinuxSandbox {
+        let mut flags =
+            libc::CLONE_NEWIPC | libc::CLONE_NEWNS | libc::CLONE_NEWPID | libc::CLONE_NEWUSER;
+        if !allow_networking {
+            flags |= libc::CLONE_NEWNET;
+        }
+
+        // Spawn sandbox init process.
+        let init_arg_raw = Box::into_raw(Box::new(init_arg));
+        let init_pid =
+            libc::clone(sandbox_init, stack_top, flags | libc::SIGCHLD, init_arg_raw as _);
+        if init_pid == -1 {
+            Err(IoError::last_os_error().into())
+        } else {
+            let mut init_arg = Box::from_raw(init_arg_raw);
+            init_arg.pid = init_pid;
+            Ok(*init_arg)
+        }
     }
 }
 
-// NOTE: Since this new process is PID 1, it will be responsible for reaping all
-// orphans. We currently do not create a reaper for these and instead leak
-// zombies, relying on the spawned process being short-lived and not spawning a
-// lot of children.
-//
-/// Sandboxing steps executed in the new process' context.
-fn post_fork(uid: u32, gid: u32, allow_networking: bool) -> io::Result<()> {
-    // Create new procfs directory.
-    let new_proc_c = CString::new("/proc").unwrap();
-    namespaces::mount_proc(&new_proc_c)?;
+/// PID 1 inside the sandbox.
+///
+/// This function is the entry point for the process which is used to launch the
+/// sandboxee and act as init system for the PID namespace.
+extern "C" fn sandbox_init(arg: *mut libc::c_void) -> libc::c_int {
+    let init_arg: Box<ProcessInitArg> = unsafe { Box::from_raw(arg as _) };
+
+    match sandbox_init_inner(*init_arg) {
+        Ok(exit_code) => exit_code,
+        Err(err) => {
+            eprintln!("sandboxing failure: {err}");
+            1
+        },
+    }
+}
 
-    // Isolate networking using a network namespace.
-    if !allow_networking {
-        namespaces::create_user_namespace(0, 0, Namespaces::NETWORK)?;
+/// PID 1 inside the sandbox.
+///
+/// Wrapper to simplify error handling.
+fn sandbox_init_inner(mut init_arg: ProcessInitArg) -> io::Result<libc::c_int> {
+    // Close all unused FDs.
+    init_arg.stdin_tx.take();
+    init_arg.stdout_rx.take();
+    init_arg.stderr_rx.take();
+
+    // Hook up stdio to parent process.
+    if let Some(stdin_pipe) = &mut init_arg.stdin_rx {
+        rustix::stdio::dup2_stdin(stdin_pipe)?;
+    }
+    if let Some(stdout_pipe) = &init_arg.stdout_tx {
+        rustix::stdio::dup2_stdout(stdout_pipe)?;
     }
+    if let Some(stderr_pipe) = &init_arg.stderr_tx {
+        rustix::stdio::dup2_stderr(stderr_pipe)?;
+    }
+
+    // Map root UID and GID.
+    namespaces::map_ids(init_arg.parent_euid.as_raw(), init_arg.parent_egid.as_raw(), 0, 0)?;
 
-    // Drop root user mapping and ensure abstract namespace is cleared.
-    namespaces::create_user_namespace(uid, gid, Namespaces::empty())?;
+    // Remove environment variables.
+    if !init_arg.full_env {
+        crate::restrict_env_variables(&init_arg.env_exceptions);
+    }
```

Comment:
I think we still need to do this **before** creating the PID namespace to avoid [leaking environment variables][1]

[1]: https://github.com/phylum-dev/birdcage/security/advisories/GHSA-wj7f-468m-6mv8

Otherwise, the sandboxed child can just read `/proc/1/environ` to recover the original values, right?

Reply:
> cat: /proc/1/environ: Permission denied

While this isn't an issue and our environment variable cleaning isn't ideal anyway (see issue in our backlog), I've still decided to move it to the parent just to be safe. Technically this should go away in the future anyway, so I don't mind affecting the hosting process for a bit longer.